Add WER metric

nikolasavic3 opened the pull request:
Add Word Error Rate (WER) metric 
Fixes #28

nikolasavic3 commented on src/metrax/metrics.py:
```diff
@@ -795,3 +795,140 @@ def merge(self, other: 'Perplexity') -> 'Perplexity':
+      return m
+
+    # Create distance matrix
+    distance_matrix = [[0 for _ in range(n+1)] for _ in range(m+1)]
```
I implemented the Levenshtein distance calculation using pure Python rather than JAX operations because the distance calculation involves updating the matrix in-place, and if we use JAX, that would create a new copy of the array for every update. I also believe that typical WER use cases have relatively short sequences, so the performance difference is negligible. Only the final results are converted to JAX arrays, which follows the pattern used in other metrics in the library.

jshin1394 replied:
sounds good to me, thank you!

jshin1394 commented on src/metrax/metrics_test.py:
```diff
@@ -320,6 +320,148 @@ def test_rsquared(self, y_true, y_pred, sample_weights):
+          np.random.randint(2, size=(2, 5, 10)).astype(np.float32),
+      ),
+  )
+  def test_perplexity(self, y_true, y_pred, sample_weights):
```
not sure why this shows up as part of this PR. This code already exists...?

jshin1394 commented on src/metrax/metrics_test.py:
```diff
@@ -320,6 +320,148 @@ def test_rsquared(self, y_true, y_pred, sample_weights):
+    )
+
 
 if __name__ == '__main__':
```
could you try syncing your workspace seems like some of more recent updates are not reflected.

jshin1394 commented on src/metrax/metrics_test.py:
```diff
@@ -320,6 +320,148 @@ def test_rsquared(self, y_true, y_pred, sample_weights):
+        atol=1e-05,
+    )
+
+  def test_wer_empty(self):
```
let's move all *_empty tests to the beginning of the file.

jshin1394 commented on src/metrax/metrics_test.py:
```diff
@@ -320,6 +320,148 @@ def test_rsquared(self, y_true, y_pred, sample_weights):
+
+    np.testing.assert_allclose(
+        metric.compute(),
+        jnp.array(3/18, dtype=jnp.float32),
```
can we test against the result gained from using a different known library? for instance, we can compare against torchmetrics https://pytorch.org/torcheval/main/generated/torcheval.metrics.WordErrorRate.html

nikolasavic3 replied:
Yes, that's a good idea. Maybe in the future, we could open a PR for keras_hub to implement the WER metric there so we have less imports? It already has ROUGE and BLEU. Or maybe we could use a simpler library like https://github.com/jitsi/jiwer.

nikolasavic3 replied:
the test now use [keras_nlp.metrics.EditDistance](https://keras.io/api/keras_nlp/metrics/edit_distance/) with `normalize=True` which is the same as computing WER, as mentioned in the documentation.

jshin1394 replied:
perfect. Thank you so much!

jshin1394 commented on src/metrax/metrics.py:
```diff
@@ -536,4 +536,141 @@ def compute(self) -> jax.Array:
+    total_edit_distance = 0
+    total_reference_length = 0
+
+    for pred, ref in zip(predictions, references):
```
should we add a TODO to use a more involved tokenizer in the future?

nikolasavic3 replied:
That's a great point! To keep things in line with the single responsibility principle, I think we should remove the tokenization logic entirely and just accept pre-tokenized lists of strings as input, leaving the tokenizer implementation to the one who calls the metric.

jshin1394 replied:
Hi Nikola! I agree that single responsibility principle is important. However, considering that it is more common for these types of metrics to take in a whole sentence, (for instance, this [torchMetrics WER](https://pytorch.org/torcheval/main/generated/torcheval.metrics.WordErrorRate.html)) let's keep the implementation such that we can take in sentences as input.

jshin1394 commented:
Thank you Nikola for your PR! Left some comments. Feel free to sync the workspace as well! :)

jshin1394 commented on src/metrax/metrics.py:
```diff
@@ -795,3 +795,137 @@ def merge(self, other: 'Perplexity') -> 'Perplexity':
+
+
+@flax.struct.dataclass
+class WER(clu_metrics.Metric):
```
should your metric inherit from clu_metrics.Average?

nikolasavic3 replied:
What do you think about it?

jshin1394 replied:
Hi Nikola! We definitely want to use these metrics in distributed settings. However, I wanted to point out that clu_metrics.Average also inherits from the clu_metrics.Metric class. Therefore, inheriting from clu_metrics.Average should be just as upgradable as clu_metrics.Metric. Please let me know if I am missing anything. Otherwise, +1 to using Average as base class! :)

nikolasavic3 replied:
You’re right, the class now inherits from Average.

nikolasavic3 commented:
Thank you Jiwon for reviewing my PR!

jshin1394 commented:
Thank you Nikola! LGTM given all the comments are resolved.

jshin1394 commented:
Hi nikola, a bunch of PRs were submitted to split the metrics.py file into multiple modules. I think the WER class can be placed under the new `nlp_metrics.py`module.

jshin1394 commented on src/metrax/nlp_metrics.py:
```diff
@@ -114,3 +117,126 @@ def merge(self, other: 'Perplexity') -> 'Perplexity':
+  def total_reference_length(self):
+    return self.count
+
+  @classmethod
```
no need to override this function as it already exists in `Average`

jshin1394 commented on src/metrax/nlp_metrics.py:
```diff
@@ -114,3 +117,126 @@ def merge(self, other: 'Perplexity') -> 'Perplexity':
+
+    return distance_matrix[m][n]
+
+  def merge(self, other: 'WER') -> 'WER':
```
no need to override this function anymore as it exists in `Average`

jshin1394 commented on src/metrax/nlp_metrics.py:
```diff
@@ -114,3 +117,126 @@ def merge(self, other: 'Perplexity') -> 'Perplexity':
+        count=self.count + other.count,
+    )
+
+  def compute(self) -> jax.Array:
```
no need to override this function as it exists in `Average`

jshin1394 commented on src/metrax/nlp_metrics.py:
```diff
@@ -114,3 +117,126 @@ def merge(self, other: 'Perplexity') -> 'Perplexity':
+      total_edit_distance: Sum of edit distances across all samples.
+      total_reference_length: Sum of reference lengths across all samples.
+  """
+  @property
```
these can be removed as well since

jshin1394 commented:
Thank you so much nikola! Almost there :) I left some comments so that we can 1) remove functions that already exist as part of [`Average`](https://github.com/google/CommonLoopUtils/blob/main/clu/metrics.py#L770) and 2) the metric can take in sentence as input following other popular metric libraries such as [torchMetrics](https://pytorch.org/torcheval/main/generated/torcheval.metrics.WordErrorRate.html).

Thank you again Nikola!

nikolasavic3 commented:
Thank you for the detailed feedback, Jiwon! I've removed all the redundant function overrides (merge, compute, total_reference_length) and updated the implementation to accept untokenized strings.

jshin1394 commented:
Thank you so much Nikola! We really appreciate your contribution to the Metrax codebase! :)
